File: 141. Linked List Cycle.cpp
/*
Given a linked list, determine if it has a cycle in it.

Follow up:
Can you solve it without using extra space?
*/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if (head == nullptr) return head;
        ListNode *p1 = head, *p2 = head;
        while (p1 != nullptr) {
            if (p1->next == nullptr) return false;
            else p1 = p1->next->next;
            p2 = p2->next;
            if (p1 == p2) return true;
        }
        return false;
    }
};

File: Linked List/141. Linked List Cycle.py
# Given a linked list, determine if it has a cycle in it.
#
# Follow up:
# Can you solve it without using extra space?

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Time: O(n), Space: O(1)

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        slowPtr = fastPtr = head
        while fastPtr and fastPtr.next:
            fastPtr = fastPtr.next.next
            slowPtr = slowPtr.next
            if slowPtr == fastPtr:
                return True
        return False
